crypto/multisig/src/lib.rs

```diff
@@ -134,47 +181,63 @@ impl<P: Default + PublicKey + Message> SignerSet<P> {
             return Err(SignatureError::new());
         }
 
-        // If the signature contains less than the threshold number of signers or more
-        // than the hardcoded limit, there's no point in trying.
-        if multi_sig.signatures.len() < self.threshold as usize
-            || multi_sig.signatures.len() > MAX_SIGNATURES
-        {
+        // Don't allow more than MAX_SIGNATURES signatures. This prevents potential
+        // abuse.
+        if multi_sig.signatures.len() > MAX_SIGNATURES {
             return Err(SignatureError::new());
         }
 
         // Sort and dedup the list of signers and signatures.
         // While the verification code below should be immune to duplicate signers or
         // signatures, the overhead of deduping them is negligible and being
         // extra-safe is a good idea.
-        let mut potential_signers = self.signers.clone();
-        potential_signers.sort();
-        potential_signers.dedup();
+        let mut potential_individual_signers = self.individual_signers.clone();
+        potential_individual_signers.sort();
+        potential_individual_signers.dedup();
+
+        let mut potential_multi_signers = self.multi_signers.clone();
+        potential_multi_signers.sort();
+        potential_multi_signers.dedup();
 
         let mut signatures = multi_sig.signatures.clone();
         signatures.sort_by(|a, b| a.as_ref().cmp(b.as_ref()));
         signatures.dedup();
 
-        // See which signatures which match signers.
-        let mut matched_signers = Vec::new();
-        for signature in signatures.iter() {
-            let matched_signer = potential_signers.iter().find_map(|signer| {
-                signer
-                    .verify(message, signature)
-                    .ok()
-                    .map(|_| signer.clone())
-            });
-            if let Some(matched_signer) = matched_signer {
-                potential_signers.retain(|signer| signer != &matched_signer);
-                matched_signers.push(matched_signer);
+        // Collect individual signer identities that signed the message, and count how
+        // many signers we successfully matched.
+        let mut matched_signer_identities = vec![];
+        let mut num_matches = 0;
+
+        for individual_signer in potential_individual_signers {
+            if signatures
+                .iter()
+                .any(|sig| individual_signer.verify(message, sig).is_ok())
+            {
+                num_matches += 1;
+                matched_signer_identities.push(individual_signer);
+            }
+        }
```

Comment:
Won't this allow one individual signer to sign multiple times and increase the count?

Reply:
How what they do that? The loop goes over the individual signers and check if there is any of the signatures provided is verifyable by their public key. Whether one or multiple signature happen to match, the code inside the `if` statement will only run once per each individual signer.